// @sap-ux/preview-middleware/dist/client/adp/quick-actions/quick-action-base.js
"use strict";

sap.ui.define([], function () {
  "use strict";

  /**
   * Base class for all  quick actions.
   */
  class QuickActionDefinitionBase {
    get id() {
      return `${this.context.key}-${this.type}`;
    }

    /**
     * Quick Actions tooltip.
     */
    get tooltip() {
      if (this.validationResult) {
        const validationErrors = this.validationResult.filter(result => result?.type === 'error');
        if (validationErrors.length > 0) {
          const error = validationErrors[0];
          return error.message;
        }
      }
      return undefined;
    }
    get isDisabled() {
      if (this.validationResult === undefined) {
        return false;
      }
      const validationErrors = this.validationResult.filter(result => result?.type === 'error');
      return validationErrors.length > 0;
    }
    get textKey() {
      return this.defaultTextKey;
    }
    constructor(type, kind, defaultTextKey, context) {
      let enablementValidators = arguments.length > 4 && arguments[4] !== undefined ? arguments[4] : [];
      this.type = type;
      this.kind = kind;
      this.defaultTextKey = defaultTextKey;
      this.context = context;
      this.enablementValidators = enablementValidators;
    }
    async runEnablementValidators() {
      this.validationResult = await Promise.all(this.enablementValidators.map(async validator => await validator.run()));
    }
  }
  var __exports = {
    __esModule: true
  };
  __exports.QuickActionDefinitionBase = QuickActionDefinitionBase;
  return __exports;
});
