// @sap-ux/preview-middleware/dist/client/cpe/context-menu-service.js
'use strict';
sap.ui.define([
    'open/ux/preview/client/thirdparty/@sap-ux-private/control-property-editor-common',
    '../adp/dialog-factory',
    '../i18n'
], function (___sap_ux_private_control_property_editor_common, ___adp_dialog_factory, ___i18n) {
    'use strict';
    const executeContextMenuAction = ___sap_ux_private_control_property_editor_common['executeContextMenuAction'];
    const requestControlContextMenu = ___sap_ux_private_control_property_editor_common['requestControlContextMenu'];
    const DialogFactory = ___adp_dialog_factory['DialogFactory'];
    const getTextBundle = ___i18n['getTextBundle'];
    class ContextMenuService {
        sendAction = () => {
        };
        constructor(rta) {
            this.rta = rta;
        }
        async init(sendAction, subscribe) {
            this.sendAction = sendAction;
            this.actionService = await this.rta.getService('action');
            const resourceBundle = await getTextBundle();
            subscribe(async action => {
                if (executeContextMenuAction.match(action)) {
                    const {actionName, controlId} = action.payload;
                    await this.actionService.execute(controlId, actionName);
                }
                if (requestControlContextMenu.pending.match(action)) {
                    const controlId = action.payload;
                    const actions = await this.actionService.get(controlId);
                    const responsePayload = {
                        controlId: controlId,
                        contextMenuItems: (actions ?? []).map(val => {
                            let enabled = val.enabled;
                            let tooltip;
                            if (!DialogFactory.canOpenDialog) {
                                enabled = false;
                                tooltip = resourceBundle.getText('ADP_QUICK_ACTION_DIALOG_OPEN_MESSAGE');
                            }
                            return {
                                id: val.id,
                                title: val.text,
                                enabled,
                                tooltip
                            };
                        })
                    };
                    const requestControlActions = requestControlContextMenu.fulfilled(responsePayload);
                    this.sendAction(requestControlActions);
                }
            });
        }
    }
    var __exports = { __esModule: true };
    __exports.ContextMenuService = ContextMenuService;
    return __exports;
});
