// @sap-ux/adp-tooling/dist/preview/change-handler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.moduleNameContentMap = void 0;
exports.tryFixChange = tryFixChange;
exports.isAddXMLChange = isAddXMLChange;
exports.isAddAnnotationChange = isAddAnnotationChange;
exports.addXmlFragment = addXmlFragment;
exports.addAnnotationFile = addAnnotationFile;
const path_1 = require("path");
const project_access_1 = require("@sap-ux/project-access");
const ejs_1 = require("ejs");
const crypto_1 = require("crypto");
const manifest_service_1 = require("../base/abap/manifest-service");
const helper_1 = require("../base/helper");
const system_access_1 = require("@sap-ux/system-access");
const odata_service_writer_1 = require("@sap-ux/odata-service-writer");
const editors_1 = require("../writer/editors");
const OBJECT_PAGE_CUSTOM_SECTION = 'OBJECT_PAGE_CUSTOM_SECTION';
const CUSTOM_ACTION = 'CUSTOM_ACTION';
const OBJECT_PAGE_HEADER_FIELD = 'OBJECT_PAGE_HEADER_FIELD';
const V2_SMART_TABLE_COLUMN = 'V2_SMART_TABLE_COLUMN';
const V2_SMART_TABLE_CELL = 'V2_SMART_TABLE_CELL';
const V4_MDC_TABLE_COLUMN = 'V4_MDC_TABLE_COLUMN';
const ANALYTICAL_TABLE_COLUMN = 'ANALYTICAL_TABLE_COLUMN';
const GRID_TREE_TABLE_COLUMN = 'GRID_TREE_TABLE_COLUMN';
const TABLE_ACTION = 'TABLE_ACTION';
const fragmentTemplateDefinitions = {
    [OBJECT_PAGE_CUSTOM_SECTION]: {
        path: 'common/op-custom-section.xml',
        getData: () => {
            const uuid = (0, crypto_1.randomBytes)(4).toString('hex');
            return {
                ids: {
                    objectPageSection: `op-section-${uuid}`,
                    objectPageSubSection: `op-subsection-${uuid}`,
                    hBox: `hbox-${uuid}`
                }
            };
        }
    },
    [CUSTOM_ACTION]: {
        path: 'common/custom-action.xml',
        getData: () => {
            const uuid = (0, crypto_1.randomBytes)(4).toString('hex');
            return {
                ids: {
                    toolbarActionButton: `btn-${uuid}`
                }
            };
        }
    },
    [TABLE_ACTION]: {
        path: 'common/v4-table-action.xml',
        getData: () => {
            const uuid = (0, crypto_1.randomBytes)(4).toString('hex');
            return {
                ids: {
                    customToolbarAction: `toolbarAction-${uuid}`,
                    customActionButton: `btn-${uuid}`
                }
            };
        }
    },
    [OBJECT_PAGE_HEADER_FIELD]: {
        path: 'common/header-field.xml',
        getData: () => {
            const uuid = (0, crypto_1.randomBytes)(4).toString('hex');
            return {
                ids: {
                    vBoxContainer: `vBox-${uuid}`,
                    label: `label-${uuid}`
                }
            };
        }
    },
    [V2_SMART_TABLE_COLUMN]: {
        path: 'v2/m-table-custom-column.xml',
        getData: (change) => {
            const uuid = (0, crypto_1.randomBytes)(4).toString('hex');
            const columnIndex = change.content.index;
            return {
                ids: {
                    column: `column-${uuid}`,
                    columnTitle: `column-title-${uuid}`,
                    customData: `custom-data-${uuid}`,
                    index: columnIndex.toFixed(0)
                }
            };
        }
    },
    [V2_SMART_TABLE_CELL]: {
        path: 'v2/m-table-custom-column-cell.xml',
        getData: () => {
            const uuid = (0, crypto_1.randomBytes)(4).toString('hex');
            return {
                ids: {
                    text: `cell-text-${uuid}`
                }
            };
        }
    },
    [V4_MDC_TABLE_COLUMN]: {
        path: 'v4/mdc-custom-column.xml',
        getData: () => {
            const uuid = (0, crypto_1.randomBytes)(4).toString('hex');
            return {
                ids: {
                    column: `column-${uuid}`,
                    text: `text-${uuid}`
                }
            };
        }
    },
    [GRID_TREE_TABLE_COLUMN]: {
        path: 'common/grid-tree-custom-column.xml',
        getData: (change) => {
            const uuid = (0, crypto_1.randomBytes)(4).toString('hex');
            const columnIndex = change.content.index;
            return {
                ids: {
                    column: `column-${uuid}`,
                    label: `label-${uuid}`,
                    text: `text-${uuid}`,
                    customData: `custom-data-${uuid}`,
                    index: columnIndex.toFixed(0)
                }
            };
        }
    },
    [ANALYTICAL_TABLE_COLUMN]: {
        path: 'common/analytical-custom-column.xml',
        getData: (change) => {
            const uuid = (0, crypto_1.randomBytes)(4).toString('hex');
            const columnIndex = change.content.index;
            return {
                ids: {
                    column: `column-${uuid}`,
                    label: `label-${uuid}`,
                    text: `text-${uuid}`,
                    customData: `custom-data-${uuid}`,
                    index: columnIndex.toFixed(0)
                }
            };
        }
    }
};
/**
 * A mapping object that defines how to extract change content data from changes based on their type.
 */
exports.moduleNameContentMap = {
    codeExt: (change) => (change.content?.codeRef ?? '').replace('.js', ''),
    addXML: (change) => change.content?.fragmentPath ?? ''
};
/**
 * Sets the moduleName property of the provided change to also support old changes with newer UI5 versions.
 *
 * @param change change to be fixed
 * @param logger logger instance
 */
/**
 * Attempts to fix a change object by setting its moduleName based on its reference and changeType
 * to also support old changes with newer UI5 versions.
 *
 * @param {CommonChangeProperties} change - The change object to be fixed.
 * @param {Logger} logger - An instance for logging warnings, errors, or informational messages.
 */
function tryFixChange(change, logger) {
    try {
        const prefix = change.reference.replace(/\./g, '/');
        change.moduleName = `${prefix}/changes/${exports.moduleNameContentMap[change.changeType](change)}`;
    }
    catch (error) {
        logger.warn('Could not fix missing module name.');
    }
}
/**
 * Determines whether a given change is of type `AddXMLChange`.
 *
 * @param {CommonChangeProperties} change - The change object to check.
 * @returns {boolean} `true` if the `changeType` is either 'addXML' or 'addXMLAtExtensionPoint',
 *          indicating the change is of type `AddXMLChange`.
 */
function isAddXMLChange(change) {
    return change.changeType === 'addXML' || change.changeType === 'addXMLAtExtensionPoint';
}
/**
 * Determines whether a given change is of type `AnnotationFileChange`.
 *
 * @param {CommonChangeProperties} change - The change object to check.
 * @returns {boolean} `true` if the `changeType` is either 'appdescr_app_addAnnotationsToOData',
 *          indicating the change is of type `AnnotationFileChange`.
 */
function isAddAnnotationChange(change) {
    return change.changeType === 'appdescr_app_addAnnotationsToOData';
}
/**
 * Asynchronously adds an XML fragment to the project if it doesn't already exist.
 *
 * @param {string} basePath - The base path of the project.
 * @param {AddXMLChange} change - The change data, including the fragment path.
 * @param {Editor} fs - The mem-fs-editor instance.
 * @param {Logger} logger - The logging instance.
 */
function addXmlFragment(basePath, change, fs, logger) {
    const { fragmentPath } = change.content;
    const fullPath = (0, path_1.join)(basePath, project_access_1.DirName.Changes, fragmentPath);
    const templateConfig = fragmentTemplateDefinitions[change.content?.templateName ?? ''];
    try {
        if (templateConfig) {
            const fragmentTemplatePath = (0, path_1.join)(__dirname, '../../templates/rta', templateConfig.path);
            const text = fs.read(fragmentTemplatePath);
            const template = (0, ejs_1.render)(text, templateConfig.getData(change));
            fs.write(fullPath, template);
        }
        else {
            // copy default fragment template
            const templateName = 'fragment.xml'; /* TemplateFileName.Fragment */
            const fragmentTemplatePath = (0, path_1.join)(__dirname, '../../templates/rta', templateName);
            fs.copy(fragmentTemplatePath, fullPath);
        }
        logger.info(`XML Fragment "${fragmentPath}" was created`);
    }
    catch (error) {
        logger.error(`Failed to create XML Fragment "${fragmentPath}": ${error}`);
    }
}
/**
 * Asynchronously adds an XML fragment to the project if it doesn't already exist.
 *
 * @param {string} basePath - The base path of the project.
 * @param {string} projectRoot - The root path of the project.
 * @param {AddXMLChange} change - The change data, including the fragment path.
 * @param {Editor} fs - The mem-fs-editor instance.
 * @param {Logger} logger - The logging instance.
 */
async function addAnnotationFile(basePath, projectRoot, change, fs, logger) {
    const { dataSourceId, annotations, dataSource } = change.content;
    const annotationDataSourceKey = annotations[0];
    const annotationUriSegments = dataSource[annotationDataSourceKey].uri.split('/');
    annotationUriSegments.shift();
    const fullPath = (0, path_1.join)(basePath, project_access_1.DirName.Changes, ...annotationUriSegments);
    try {
        const manifestService = await getManifestService(projectRoot, logger);
        const metadata = await manifestService.getDataSourceMetadata(dataSourceId);
        const datasoruces = await manifestService.getManifestDataSources();
        const namespaces = (0, odata_service_writer_1.getAnnotationNamespaces)({ metadata });
        await (0, editors_1.generateChange)(projectRoot, "appdescr_app_addAnnotationsToOData" /* ChangeType.ADD_ANNOTATIONS_TO_ODATA */, {
            annotation: {
                dataSource: dataSourceId,
                namespaces,
                serviceUrl: datasoruces[dataSourceId].uri,
                fileName: (0, path_1.basename)(dataSource[annotationDataSourceKey].uri)
            },
            variant: (0, helper_1.getVariant)(projectRoot),
            isCommand: false
        }, fs);
    }
    catch (error) {
        logger.error(`Failed to create Local Annotation File "${fullPath}": ${error}`);
        throw new Error('Failed to create Local Annotation File' + error.message);
    }
}
/**
 * Returns manifest service.
 *
 * @param {string} basePath - The base path of the project.
 * @param {Logger} logger - The logging instance.
 * @returns Promise<ManifestService>
 */
async function getManifestService(basePath, logger) {
    const variant = (0, helper_1.getVariant)(basePath);
    const { target, ignoreCertErrors = false } = await (0, helper_1.getAdpConfig)(basePath, (0, path_1.join)(basePath, project_access_1.FileName.Ui5Yaml));
    const provider = await (0, system_access_1.createAbapServiceProvider)(target, { ignoreCertErrors }, true, logger);
    return await manifest_service_1.ManifestService.initMergedManifest(provider, basePath, variant, logger);
}
