// @sap-ux/adp-tooling/dist/preview/routes-handler.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const fs = __importStar(require("fs"));
const os = __importStar(require("os"));
const path = __importStar(require("path"));
const ejs_1 = require("ejs");
const sanitize_filename_1 = __importDefault(require("sanitize-filename"));
const btp_utils_1 = require("@sap-ux/btp-utils");
const project_access_1 = require("@sap-ux/project-access");
const manifest_service_1 = require("../base/abap/manifest-service");
const helper_1 = require("../base/helper");
const system_access_1 = require("@sap-ux/system-access");
/**
 * @description Handles API Routes
 */
class RoutesHandler {
    project;
    util;
    logger;
    /**
     * Constructor taking project as input.
     *
     * @param project Reference to the root of the project
     * @param util middleware utilities provided by the UI5 CLI
     * @param logger Logger instance
     */
    constructor(project, util, logger) {
        this.project = project;
        this.util = util;
        this.logger = logger;
    }
    /**
     * Reads files from workspace by given search pattern.
     *
     * @param pattern Search pattern
     * @returns Array of files
     */
    async readAllFilesByGlob(pattern) {
        return this.project.byGlob(pattern);
    }
    /**
     * Sends response with data to the client.
     *
     * @param res Response
     * @param data Data that is sent to the client
     * @param contentType Content type, defaults to json
     */
    sendFilesResponse(res, data, contentType = 'application/json') {
        res.status(200 /* HttpStatusCodes.OK */).contentType(contentType).send(data);
    }
    /**
     * Sanitizes and handles error messages.
     *
     * @param res Response
     * @param next Next function
     * @param e Error
     * @param e.message Error message
     */
    handleErrorMessage(res, next, e) {
        const sanitizedMsg = (0, sanitize_filename_1.default)(e.message);
        this.logger.error(sanitizedMsg);
        res.status(500 /* HttpStatusCodes.INTERNAL_SERVER_ERROR */).send({ message: sanitizedMsg });
        next(e);
    }
    /**
     * Handler for reading all fragment files from the workspace.
     *
     * @param _ Request
     * @param res Response
     * @param next Next Function
     */
    handleReadAllFragments = async (_, res, next) => {
        try {
            const files = await this.readAllFilesByGlob('/**/changes/fragments/*.fragment.xml');
            const fileNames = files.map((f) => ({
                fragmentName: f.getName()
            }));
            this.sendFilesResponse(res, {
                fragments: fileNames,
                message: `${fileNames.length} fragments found in the project workspace.`
            });
            this.logger.debug(`Read fragments ${JSON.stringify(fileNames)}`);
        }
        catch (e) {
            this.handleErrorMessage(res, next, e);
        }
    };
    /**
     * Handler for reading all controller extension files from the workspace.
     *
     * @param _ Request
     * @param res Response
     * @param next Next Function
     */
    handleReadAllControllers = async (_, res, next) => {
        try {
            const files = await this.readAllFilesByGlob('/**/changes/coding/*.js');
            const fileNames = files.map((f) => ({
                controllerName: f.getName()
            }));
            this.sendFilesResponse(res, {
                controllers: fileNames,
                message: `${fileNames.length} controllers found in the project workspace.`
            });
            this.logger.debug(`Read controllers ${JSON.stringify(fileNames)}`);
        }
        catch (e) {
            this.handleErrorMessage(res, next, e);
        }
    };
    /**
     * Handler for retrieving existing controller extension data from the workspace.
     *
     * @param req Request
     * @param res Response
     * @param next Next Function
     */
    handleGetControllerExtensionData = async (req, res, next) => {
        try {
            const params = req.params;
            const controllerName = (0, sanitize_filename_1.default)(params.controllerName);
            const codeExtFiles = await this.readAllFilesByGlob('/**/changes/*_codeExt.change');
            let controllerPathFromRoot = '';
            let controllerExists = false;
            let controllerPath = '';
            let changeFilePath = '';
            const project = this.util.getProject();
            const sourcePath = project.getSourcePath();
            const projectName = project.getName();
            const getPath = (projectPath, fileName, folder = project_access_1.DirName.Coding) => path.join(projectPath, project_access_1.DirName.Changes, folder, fileName).split(path.sep).join(path.posix.sep);
            for (const file of codeExtFiles) {
                const fileStr = await file.getString();
                const change = JSON.parse(fileStr);
                if (change.selector.controllerName === controllerName) {
                    const fileName = change.content.codeRef.replace('coding/', '');
                    controllerPath = getPath(sourcePath, fileName);
                    controllerPathFromRoot = getPath(projectName, fileName);
                    changeFilePath = getPath(projectName, file.getName(), '');
                    controllerExists = true;
                    break;
                }
            }
            if (controllerExists && !fs.existsSync(controllerPath)) {
                const errorMsg = `Please delete the change file at "${changeFilePath}" and retry creating the controller extension.`;
                this.logger.debug(errorMsg);
                res.status(404 /* HttpStatusCodes.NOT_FOUND */).send({ message: errorMsg });
                return;
            }
            const isRunningInBAS = (0, btp_utils_1.isAppStudio)();
            this.sendFilesResponse(res, {
                controllerExists,
                controllerPath: os.platform() === 'win32' ? `/${controllerPath}` : controllerPath,
                controllerPathFromRoot,
                isRunningInBAS
            });
            this.logger.debug(controllerExists
                ? `Controller exists at '${controllerPath}'`
                : `Controller with controllerName '${controllerName}' does not exist`);
        }
        catch (e) {
            this.handleErrorMessage(res, next, e);
        }
    };
    /**
     * Handler for writing a controller extension file to the workspace.
     *
     * @param req Request
     * @param res Response
     * @param next Next Function
     */
    handleWriteControllerExt = async (req, res, next) => {
        try {
            const data = req.body;
            const controllerExtName = (0, sanitize_filename_1.default)(data.controllerName);
            const projectId = data.projectId;
            const sourcePath = this.util.getProject().getSourcePath();
            if (!controllerExtName) {
                res.status(400 /* HttpStatusCodes.BAD_REQUEST */).send('Controller extension name was not provided!');
                this.logger.debug('Bad request. Controller extension name was not provided!');
                return;
            }
            const fullPath = path.join(sourcePath, project_access_1.DirName.Changes, project_access_1.DirName.Coding);
            const filePath = path.join(fullPath, `${controllerExtName}.js`);
            if (!fs.existsSync(fullPath)) {
                fs.mkdirSync(fullPath, { recursive: true });
            }
            if (fs.existsSync(filePath)) {
                res.status(409 /* HttpStatusCodes.CONFLICT */).send(`Controller extension with name "${controllerExtName}" already exists`);
                this.logger.debug(`Controller extension with name "${controllerExtName}" already exists`);
                return;
            }
            const controllerExtPath = `${projectId}.${controllerExtName}`;
            const controllerTemplateFilePath = path.join(__dirname, '../../templates/rta', "controller.ejs" /* TemplateFileName.Controller */);
            (0, ejs_1.renderFile)(controllerTemplateFilePath, { controllerExtPath }, {}, (err, str) => {
                if (err) {
                    throw new Error('Error rendering template: ' + err.message);
                }
                fs.writeFileSync(filePath, str, { encoding: 'utf8' });
            });
            const message = 'Controller extension created!';
            res.status(201 /* HttpStatusCodes.CREATED */).send(message);
            this.logger.debug(`Controller extension with name "${controllerExtName}" was created`);
        }
        catch (e) {
            const sanitizedMsg = (0, sanitize_filename_1.default)(e.message);
            this.logger.error(sanitizedMsg);
            res.status(500 /* HttpStatusCodes.INTERNAL_SERVER_ERROR */).send(sanitizedMsg);
            next(e);
        }
    };
    /**
     * Handler for mapping annotation files with datasoruce.
     *
     * @param _req Request
     * @param res Response
     * @param next Next Function
     */
    handleGetAllAnnotationFilesMappedByDataSource = async (_req, res, next) => {
        try {
            const isRunningInBAS = (0, btp_utils_1.isAppStudio)();
            const manifestService = await this.getManifestService();
            const dataSources = manifestService.getManifestDataSources();
            const apiResponse = {
                isRunningInBAS,
                annotationDataSourceMap: {}
            };
            for (const dataSourceId in dataSources) {
                if (dataSources[dataSourceId].type === 'OData') {
                    apiResponse.annotationDataSourceMap[dataSourceId] = {
                        annotationDetails: {
                            annotationExistsInWS: false
                        },
                        serviceUrl: dataSources[dataSourceId].uri
                    };
                }
                this.fillAnnotationDataSourceMap(dataSources, dataSourceId, apiResponse.annotationDataSourceMap);
            }
            this.sendFilesResponse(res, apiResponse);
        }
        catch (e) {
            this.handleErrorMessage(res, next, e);
        }
    };
    /**
     * Add local annotation details to api response.
     *
     * @param dataSources DataSources
     * @param dataSourceId string
     * @param apiResponse AnnotationDataSourceMap
     */
    fillAnnotationDataSourceMap(dataSources, dataSourceId, apiResponse) {
        const project = this.util.getProject();
        const getPath = (projectPath, relativePath) => path.join(projectPath, relativePath).split(path.sep).join(path.posix.sep);
        const annotations = [...(dataSources[dataSourceId].settings?.annotations ?? [])].reverse();
        for (const annotation of annotations) {
            const annotationSetting = dataSources[annotation];
            if (annotationSetting.type === 'ODataAnnotation') {
                const ui5NamespaceUri = `ui5://${project.getNamespace()}`;
                if (annotationSetting.uri.startsWith(ui5NamespaceUri)) {
                    const localAnnotationUri = annotationSetting.uri.replace(ui5NamespaceUri, '');
                    const annotationPath = getPath(project.getSourcePath(), localAnnotationUri);
                    const annotationPathFromRoot = getPath(project.getName(), localAnnotationUri);
                    const annotationExists = fs.existsSync(annotationPath);
                    apiResponse[dataSourceId].annotationDetails = {
                        fileName: path.parse(localAnnotationUri).base,
                        annotationPath: os.platform() === 'win32' ? `/${annotationPath}` : annotationPath,
                        annotationPathFromRoot,
                        annotationExistsInWS: annotationExists
                    };
                }
                if (apiResponse[dataSourceId].annotationDetails.annotationExistsInWS) {
                    break;
                }
            }
        }
    }
    /**
     * Returns manifest service.
     *
     * @returns Promise<ManifestService>
     */
    async getManifestService() {
        const project = this.util.getProject();
        const basePath = project.getRootPath();
        const variant = (0, helper_1.getVariant)(basePath);
        const { target, ignoreCertErrors = false } = await (0, helper_1.getAdpConfig)(basePath, path.join(basePath, project_access_1.FileName.Ui5Yaml));
        const provider = await (0, system_access_1.createAbapServiceProvider)(target, { ignoreCertErrors }, true, this.logger);
        return await manifest_service_1.ManifestService.initMergedManifest(provider, basePath, variant, this.logger);
    }
}
exports.default = RoutesHandler;
