// @sap-ux/preview-middleware/dist/client/adp/quick-actions/fe-v4/utils.js
"use strict";

sap.ui.define(["../../../utils/core", "sap/ui/rta/command/CommandFactory", "../../../utils/fe-v4"], function (_____utils_core, CommandFactory, _____utils_fe_v4) {
  "use strict";

  const getControlById = _____utils_core["getControlById"];
  const getV4AppComponent = _____utils_fe_v4["getV4AppComponent"];
  const getPageName = _____utils_fe_v4["getPageName"];
  const getReference = _____utils_fe_v4["getReference"];
  async function executeToggleAction(context, isButtonEnabled, controlType, propertyPath) {
    const controls = context.controlIndex[controlType] ?? [];
    const control = controls[0];
    if (control) {
      const modifiedControl = getControlById(control.controlId);
      if (!modifiedControl) {
        return [];
      }
      const {
        flexSettings
      } = context;
      const parent = modifiedControl.getParent();
      if (!parent) {
        return [];
      }
      const modifiedValue = {
        reference: getReference(modifiedControl),
        appComponent: getV4AppComponent(modifiedControl),
        changeType: 'appdescr_fe_changePageConfiguration',
        parameters: {
          page: getPageName(parent),
          entityPropertyChange: {
            propertyPath: propertyPath,
            propertyValue: !isButtonEnabled,
            operation: 'UPSERT'
          }
        }
      };
      const command = await CommandFactory.getCommandFor(modifiedControl, 'appDescriptor', modifiedValue, null, flexSettings);
      return [command];
    }
    return [];
  }
  var __exports = {
    __esModule: true
  };
  __exports.executeToggleAction = executeToggleAction;
  return __exports;
});
