// @sap-ux/btp-utils/dist/app-studio.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BAS_DEST_INSTANCE_CRED_HEADER = void 0;
exports.isAppStudio = isAppStudio;
exports.getAppStudioProxyURL = getAppStudioProxyURL;
exports.getAppStudioBaseURL = getAppStudioBaseURL;
exports.getCredentialsForDestinationService = getCredentialsForDestinationService;
exports.getDestinationUrlForAppStudio = getDestinationUrlForAppStudio;
exports.listDestinations = listDestinations;
exports.exposePort = exposePort;
const axios_1 = __importDefault(require("axios"));
const cf_tools_1 = require("@sap/cf-tools");
const app_studio_env_1 = require("./app-studio.env");
const destination_1 = require("./destination");
/**
 * HTTP header that is to be used for encoded credentials when communicating with a destination service instance.
 */
exports.BAS_DEST_INSTANCE_CRED_HEADER = 'bas-destination-instance-cred';
/**
 * Check if this is exectued in SAP Business Application Studio.
 *
 * @returns true if yes
 */
function isAppStudio() {
    return !!process.env[app_studio_env_1.ENV.H2O_URL];
}
/**
 * Read and return the BAS proxy url.
 *
 * @returns the proxy url or undefined if called outside of BAS.
 */
function getAppStudioProxyURL() {
    return process.env[app_studio_env_1.ENV.PROXY_URL];
}
/**
 * Read and return the BAS base url.
 *
 * @returns the base url or undefined if called outside of BAS.
 */
function getAppStudioBaseURL() {
    return process.env[app_studio_env_1.ENV.H2O_URL];
}
/**
 * Asynchronously creates a base64 encoded credentials for the given destination service instance based on the client information fetched from BTP.
 *
 * @param instance name/id of the destination service instance
 * @returns the base64 encoded user
 */
async function getCredentialsForDestinationService(instance) {
    try {
        const serviceInfo = await (0, cf_tools_1.cfGetInstanceKeyParameters)(instance);
        if (!serviceInfo) {
            throw new Error(`No destination instance ${instance} found`);
        }
        const serviceCredentials = serviceInfo.credentials;
        if (!serviceCredentials) {
            throw new Error(`No credentials for destination instance ${instance} found`);
        }
        const clientId = serviceCredentials.uaa?.clientid || serviceCredentials.clientid;
        const clientSecret = serviceCredentials.uaa?.clientsecret || serviceCredentials.clientsecret;
        return Buffer.from(`${encodeURIComponent(clientId)}:${encodeURIComponent(clientSecret)}`).toString('base64');
    }
    catch (error) {
        throw new Error(`An error occurred while retrieving service key for the destination instance ${instance}: ${error}`);
    }
}
/**
 * Returns a url for AppStudio for the given url with the given destination.
 *
 * @param name name of the destination
 * @param path optional path
 * @returns destination url working in BAS
 */
function getDestinationUrlForAppStudio(name, path) {
    const origin = `https://${name}.dest`;
    return path && path.length > 1 ? new URL(path, origin).toString() : origin;
}
/**
 * Helper function to strip `-api` from the host name.
 *
 * @param host -
 * @returns
 */
function stripS4HCApiHost(host) {
    const [first, ...rest] = host.split('.');
    return [first.replace(/-api$/, ''), ...rest].join('.');
}
/**
 * Get a list of available destinations in SAP Business Application Studio.
 *
 * @param options - options for the destinations
 * @returns the list of destinations
 */
async function listDestinations(options) {
    const destinations = {};
    await axios_1.default.get('/reload', { baseURL: process.env[app_studio_env_1.ENV.PROXY_URL] });
    const response = await axios_1.default.get('/api/listDestinations', { baseURL: process.env[app_studio_env_1.ENV.H2O_URL] });
    const list = Array.isArray(response.data) ? response.data : [];
    list.forEach((destination) => {
        if (options?.stripS4HCApiHosts && (0, destination_1.isS4HC)(destination)) {
            destination.Host = stripS4HCApiHost(destination.Host);
        }
        if (destination.WebIDEEnabled) {
            destinations[destination.Name] = destination;
        }
    });
    return destinations;
}
/**
 * Exposes port in SAP Business Application Studio.
 *
 * @param port Port that needs to be exposed
 * @param logger Logger
 * @returns url on which the port is exposed
 */
async function exposePort(port, logger) {
    try {
        const response = await axios_1.default.get(`http://localhost:3001/AppStudio/api/getHostByPort?port=${port}`);
        return `${response.data.result}`;
    }
    catch (error) {
        logger?.error(`Port ${port} was not exposed!`);
        return '';
    }
}
